### PR TITLE
Add configuration type to prevent archetype slot filling in operational templates

### PR DESCRIPTION
**Pull request by VeraPrinsen**
…al templates

### CODE REVIEW

**Review by pieterbos — COMMENTED** [review]
Is this only to enable creating an operational template when an included archetype cannot be found, or cannot be flattened? Or do you really need the form without slots filled?

In the first case, should this be the solution, or should it be an option to just not fill the slots that cannot be found, so a form of graceful degradation?

File: tools/src/main/java/com/nedap/archie/flattener/FlattenerConfiguration.java [hunk 1]
```diff
@@ -52,6 +52,11 @@
      */
     private boolean fillEmptyOccurrences = true;
 
+    /**
+     * Only for Operational templates: fill archetype slots. Default to true
+     */
+    private boolean fillSlots = true;
```
Comment (pieterbos) [inline tools/src/main/java/com/nedap/archie/flattener/FlattenerConfiguration.java:58]: I think the correct term is not fill slots - it is 'replace archetype roots with actual archetypes'.

File: tools/src/main/java/com/nedap/archie/flattener/OperationalTemplateCreator.java [hunk 2]
```diff
@@ -50,6 +50,10 @@ public static void overrideArchetypeId(Archetype result, Archetype override) {
     }
 
     public void fillSlots(OperationalTemplate archetype) { //should this be OperationalTemplate?
+        if(!getConfig().isFillSlots()) {
```
Comment (pieterbos) [inline tools/src/main/java/com/nedap/archie/flattener/OperationalTemplateCreator.java:53]: This skips:
1. removing closed archetype slots
2. replacing archetype roots with their inline copy
3. replacing complex object proxies with a copy of the referenced structure.

I think you only want to skip 2., and not the rest? 

If not, your configuration parameter needs an entirely different description, and then I would suggest three separate parameters.

**Review by pieterbos — CHANGES_REQUESTED** [review]

**Comment by VeraPrinsen** [comment]
> Is this only to enable creating an operational template when an included archetype cannot be found, or cannot be flattened? Or do you really need the form without slots filled?
> 
> In the first case, should this be the solution, or should it be an option to just not fill the slots that cannot be found, so a form of graceful degradation?

I think both options should be possible and then it is up to the user of archie to use this. Or should archie automatically provide the operational template without archetype roots filled in if it is not possible? I don't think that is desirable.

**Comment by pieterbos** [comment]
> I think both options should be possible and then it is up to the user of archie to use this. Or should archie automatically provide the operational template without archetype roots filled in if it is not possible? I don't think that is desirable.

I think the behaviour of not giving an error when an archetype is missing, but just keeping the archetype root, could be a configurable option. Then you would get an operational template that is as good as possible given the provided information. That would not be a good idea for a production system, but could be useful for modelling tools. Just as the configuration option for this PR is not useful for production systems.

I wonder if this configurable graceful degradation is what you need for the modelling application you are working on, or actually if the configuration parameter from this PR is what you need.

This PR could still be usable, although this is mostly equal to the flattened form with the use_nodes replaced.

**Pull request closed without merging**